nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/iot/PutAWSIoT.java

```diff
@@ -0,0 +1,141 @@
+                    "a new one, but one that has been retained by the MQTT server.")
+            .required(true)
+            .defaultValue(PROP_DEFAULT_RETAINED.toString())
+            .addValidator(StandardValidators.BOOLEAN_VALIDATOR)
+            .build();
+
+    public static final List<PropertyDescriptor> properties = Collections.unmodifiableList(
+            Arrays.asList(
+                    PROP_QOS,
+                    PROP_TOPIC,
+                    PROP_RETAINED,
+                    PROP_ENDPOINT,
+                    PROP_KEEPALIVE,
+                    PROP_CLIENT,
+                    AWS_CREDENTIALS_PROVIDER_SERVICE,
+                    TIMEOUT,
+                    REGION));
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return properties;
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return Collections.singleton(REL_SUCCESS);
+    }
+
+    @OnScheduled
+    public void onScheduled(final ProcessContext context) {
+        shouldRetain = context.getProperty(PROP_RETAINED).isSet() ? context.getProperty(PROP_RETAINED).asBoolean() : PROP_DEFAULT_RETAINED;
+        init(context);
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) {
+        // check if MQTT-connection is about to expire
+        if (isConnectionAboutToExpire()) {
+            // renew connection
+            mqttClient = connect(context);
+        }
+        // get flowfile
+        FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
+        Map<String, String> attributes = flowFile.getAttributes();
+        // if provided override MQTT configuration with values from the corresponding message attributes
+        String topic = attributes.containsKey(ATTR_NAME_TOPIC) ? attributes.get(ATTR_NAME_TOPIC) : awsTopic;
+        Integer qos = attributes.containsKey(ATTR_NAME_QOS) ? Integer.parseInt(attributes.get(ATTR_NAME_QOS)) : awsQos;
+        Boolean retained = attributes.containsKey(ATTR_NAME_RETAINED) ? Boolean.parseBoolean(attributes.get(ATTR_NAME_RETAINED)) : shouldRetain;
+        // get message content
+        final ByteArrayOutputStream fileContentStream = new ByteArrayOutputStream();
+        session.exportTo(flowFile, fileContentStream);
+
+        try {
+            // publish messages to mqtt-topic(s)
+            mqttClient.publish(topic, fileContentStream.toByteArray(), qos, retained);
+            session.transfer(flowFile, REL_SUCCESS);
+            session.getProvenanceReporter().send(flowFile, awsEndpoint + "(" + awsClientId + ")");
+        } catch (MqttException e) {
+            getLogger().error("Error while initially subscribing to topics with client " + mqttClient.getClientId() + " caused by " + e.getMessage());
+            flowFile = session.putAttribute(flowFile, ATTR_NAME_EXCEPTION, e.getMessage());
```

Comment:
I swear I saw it somewhere and thought it is a practice of exposing errors in NiFi. Actually this is not a good pattern at all. Removed it.